### PR TITLE
BUG: Fix advanced object non-array value assignment

### PR DESCRIPTION
**Pull request by seberg**
If the value array in an advanced indexing assignment was a non-array
and the array being assigned to is an object array, a temporary
array is created to make the object assignment work well like
slicing in general.
However, this array was created with the swapping of the mapiter
object. So we have to swap it back as if it was an array being
subscripted with MapIterSwapAxes.

---

This is seems like a pretty bad bug, though then again I am impressed nobody noticed, since most of the time it should just make code fail....

### CODE REVIEW

**Comment by seberg** [comment]
Hmmm, I tried to make it a common merge base, seems that was not successful?! Anyone a quick idea why?

**Comment by seberg** [comment]
Ah, nvm there is a fallback added since the splitting of the two...

**Comment by seberg** [comment]
Ok, we will have to cherry pick, unless someone has a nice idea how to do it otherwise.

**Comment by jaimefrio** [comment]
Does this fix #5710 too?

**Comment by seberg** [comment]
Nope, thats unrelated (well, obviously it has the same reason, and that is when I realized this might be buggy).

File: numpy/core/tests/test_indexing.py [hunk 1]
```diff
@@ -547,6 +547,17 @@ def test_object_assign(self):
         b[[0], :3] = [[1, (1,2), 3]]
         assert_array_equal(a, b[0])
 
+        # Check that swapping of axes works
+        b = b.T
+        b[:3, [0]] = [[1], [(1,2)], [3]]
+        assert_array_equal(a, b[:, 0])
+
+        # Another test for the memory order of the subspace
+        # (though it should not matter)
```
Comment (charris) [inline numpy/core/tests/test_indexing.py]: Looks like the test in the assignments is that no exception is raised? That should explicitly tested if so.

**Comment by charris** [comment]
LGTM modulo test clarification.

File: numpy/core/tests/test_indexing.py (hunk 1 above)
Comment (seberg) [inline numpy/core/tests/test_indexing.py]: Yeah, it just tests that, though actually maybe I could also test if the result is actually correct anyway. But how would you explictly test it?

Comment (charris) [inline numpy/core/tests/test_indexing.py]: For the exceptions, I'd just put it in a `try ... except` construct and raise `AssertionError` on exception. That makes the intent clear, and a comment that the assignment should not raise an exception would emphasize that.

Comment (seberg) [inline numpy/core/tests/test_indexing.py]: Comparing the output the expected output now. Added the try/except for the first part. The second part actually never failed with an error, but checking the correctness is good anyway (though I admit, the more likely thing here is probably an error then incorrect result).

File: numpy/core/tests/test_indexing.py [hunk 2]
```diff
@@ -547,6 +547,25 @@ def test_object_assign(self):
         b[[0], :3] = [[1, (1,2), 3]]
         assert_array_equal(a, b[0])
 
+        # Check that swapping of axes works.
+        # This used to throw an error.
+        b = b.T
+        try:
+            b[:3, [0]] = [[1], [(1,2)], [3]]
+        except ValueError:
+            raise AssertionError("Assignment is possible")
```
Comment (charris) [inline numpy/core/tests/test_indexing.py]: Maybe `Assignment fails`?

Comment (pv) [inline numpy/core/tests/test_indexing.py]: Typically, we do not convert exceptions to AssertionErrors in tests. So
just removing the `try: except:` block would also be OK.

Comment (charris) [inline numpy/core/tests/test_indexing.py]: I would be OK with that also as long as the comment remains.

Comment (seberg) [inline numpy/core/tests/test_indexing.py]: OK, just removed it and made the comment a bit more clear + mentioning of the gh-issue.

**Comment by charris** [comment]
Thanks Sebastian.

**Pull request merged**